pkg/oidc/oidc.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 The Lokomotive Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oidc
+
+import (
+	"fmt"
+	"net/url"
+)
+
+const (
+	defaultClientID      = "gangway"
+	defaultUsernameClaim = "email"
+	defaultGroupsClaim   = "groups"
+
+	issuerURLKey     = "--oidc-issuer-url"
+	clientIDKey      = "--oidc-client-id"
+	usernameClaimKey = "--oidc-username-claim"
+	groupsClaimKey   = "--oidc-groups-claim"
+)
+
+// Config deals with providing OIDC related fields to the Lokomotive
+// cluster for OIDC based Authentication via Dex and Gangway.
+type Config struct {
+	IssuerURL     string `hcl:"issuer_url,optional"`
+	ClientID      string `hcl:"client_id,optional"`
+	UsernameClaim string `hcl:"username_claim,optional"`
+	GroupsClaim   string `hcl:"groups_claim,optional"`
+}
+
+//setDefaultValuesForEmptyFields sets default values for empty fields.
+func (c *Config) setDefaultValuesForEmptyFields() {
+	if c.ClientID == "" {
+		c.ClientID = defaultClientID
+	}
+
+	if c.UsernameClaim == "" {
+		c.UsernameClaim = defaultUsernameClaim
+	}
+
+	if c.GroupsClaim == "" {
+		c.GroupsClaim = defaultGroupsClaim
+	}
+}
+
+// GetConfiguredOIDCFlags get a list of oidc flags after setting default
+// values and validations.
+func (c *Config) GetConfiguredOIDCFlags() ([]string, error) {
+	// Set default values for fields empty/not passed by the user.
+	c.setDefaultValuesForEmptyFields()
+
+	if err := c.validate(); err != nil {
+		return []string{}, err
+	}
+	// initialize map
+	oidcFlags := map[string]string{}
+	oidcFlags[issuerURLKey] = c.IssuerURL
+	oidcFlags[clientIDKey] = c.ClientID
+	oidcFlags[usernameClaimKey] = c.UsernameClaim
+	oidcFlags[groupsClaimKey] = c.GroupsClaim
+
+	oidcFlagsList := []string{}
+
+	for k, v := range oidcFlags {
+		oidcFlagsList = append(oidcFlagsList, fmt.Sprintf("\"%s=%s\"", k, v))
+	}
+
+	return oidcFlagsList, nil
+}
+
+//validate validates the values of the oidc Config fields.
+func (c *Config) validate() error {
```

Comment:
I would make validate exported. This allows user to validate manually and skip error checking when calling `GetConfiguredOIDCFlags()`.

Then, validate should also call `c.setDefaultValuesForEmptyFields()`.